### ISSUE TITLE
unmount the volume failed

### ISSUE THREAD

**Issue by NickrenREN** [kubernetes/kubernetes#36983]
**Kubernetes version** (use `kubectl version`):

Client Version: version.Info{Major:"1", Minor:"2", GitVersion:"v1.2.0", GitCommit:"5e723f67f1e36d387a8a7faa6aa8a7f40cc9ca46", GitTreeState:"clean"}
Server Version: version.Info{Major:"1", Minor:"5+", GitVersion:"v1.5.0-alpha.1.859+cfbb0eb2ff6801", GitCommit:"cfbb0eb2ff680173eeaa0ce91ab42f3c4f9bbc33", GitTreeState:"clean"}

**Environment**:
- **Cloud provider or hardware configuration**: 

openstack with blockstorage cinder

- **OS** (e.g. from /etc/os-release):
- **Kernel** (e.g. `uname -a`):
- **Install tools**:
- **Others**:

**What happened**:
I have already deleted the pod .but kubelet logs as follows:
E1117 10:25:41.898748   29955 nestedpendingoperations.go:253] Operation for "\"kubernetes.io/cinder/ubuntu-persistent-storage\" (\"4389375b-acab-11e6-ac07-fa163e411bde\")" failed. No retries permitted until 2016-11-17 10:27:41.898715848 +0000 UTC (durationBeforeRetry 2m0s). Error: UnmountVolume.TearDown failed for volume "kubernetes.io/cinder/ubuntu-persistent-storage" (volume.spec.Name: "ubuntu-persistent-storage") pod "4389375b-acab-11e6-ac07-fa163e411bde" (UID: "4389375b-acab-11e6-ac07-fa163e411bde") with: remove /var/lib/kubelet/pods/4389375b-acab-11e6-ac07-fa163e411bde/volumes/kubernetes.io~cinder/ubuntu-persistent-storage: device or resource busy

**What you expected to happen**:

unmount the volume successfully

**How to reproduce it** (as minimally and precisely as possible):

kubernetes run with blockstorage cinder

**Anything else do we need to know**:

**Comment by prg3** [comment]
Does lsof give you any indication of what files may remain open after the pod is down?

**Comment by xiangpengzhao** [comment]
I also encounter similar issue, kubernetes produces tons of log such as:
```
E0313 01:26:46.592912   12770 nestedpendingoperations.go:262] Operation for "\"kubernetes.io/empty-dir/26b15e34-0565-11e7-98db-fa163e09f30e-app-logs.deleting~126800867.deleting~198252498.deleting~495025749.deleting~948474367.deleting~535829293\" (\"26b15e34-0565-11e7-98db-fa163e09f30e\")" failed. No retries permitted until 2017-03-13 01:26:47.092826492 -0700 PDT (durationBeforeRetry 500ms). Error: UnmountVolume.TearDown failed for volume "kubernetes.io/empty-dir/26b15e34-0565-11e7-98db-fa163e09f30e-app-logs.deleting~126800867.deleting~198252498.deleting~495025749.deleting~948474367.deleting~535829293" (volume.spec.Name: "app-logs.deleting~126800867.deleting~198252498.deleting~495025749.deleting~948474367.deleting~535829293") pod "26b15e34-0565-11e7-98db-fa163e09f30e" (UID: "26b15e34-0565-11e7-98db-fa163e09f30e") with: rename /var/lib/kubelet/pods/26b15e34-0565-11e7-98db-fa163e09f30e/volumes/kubernetes.io~empty-dir/app-logs.deleting~126800867.deleting~198252498.deleting~495025749.deleting~948474367.deleting~535829293 /var/lib/kubelet/pods/26b15e34-0565-11e7-98db-fa163e09f30e/volumes/kubernetes.io~empty-dir/app-logs.deleting~126800867.deleting~198252498.deleting~495025749.deleting~948474367.deleting~535829293.deleting~129755443: file exists
E0313 01:26:46.733737   12770 nestedpendingoperations.go:262] Operation for "\"kubernetes.io/empty-dir/26b15e34-0565-11e7-98db-fa163e09f30e-app-logs.deleting~210568084.deleting~007091719.deleting~563227903.deleting~203159535.deleting~551798173\" (\"26b15e34-0565-11e7-98db-fa163e09f30e\")" failed. No retries permitted until 2017-03-13 01:26:47.233660496 -0700 PDT (durationBeforeRetry 500ms). Error: UnmountVolume.TearDown failed for volume "kubernetes.io/empty-dir/26b15e34-0565-11e7-98db-fa163e09f30e-app-logs.deleting~210568084.deleting~007091719.deleting~563227903.deleting~203159535.deleting~551798173" (volume.spec.Name: "app-logs.deleting~210568084.deleting~007091719.deleting~563227903.deleting~203159535.deleting~551798173") pod "26b15e34-0565-11e7-98db-fa163e09f30e" (UID: "26b15e34-0565-11e7-98db-fa163e09f30e") with: rename /var/lib/kubelet/pods/26b15e34-0565-11e7-98db-fa163e09f30e/volumes/kubernetes.io~empty-dir/app-logs.deleting~210568084.deleting~007091719.deleting~563227903.deleting~203159535.deleting~551798173 /var/lib/kubelet/pods/26b15e34-0565-11e7-98db-fa163e09f30e/volumes/kubernetes.io~empty-dir/app-logs.deleting~210568084.deleting~007091719.deleting~563227903.deleting~203159535.deleting~551798173.deleting~284576744: file exists
```

and there are tons of directories in directory `/var/lib/kubelet/pods/26b15e34-0565-11e7-98db-fa163e09f30e/volumes/kubernetes.io~empty-dir`
```
app-logs.deleting~895432393.deleting~957086549.deleting~943289565.deleting~124936451.deleting~004071477
app-logs.deleting~895432393.deleting~957086549.deleting~943289565.deleting~124936451.deleting~004071477.deleting~428510874
app-logs.deleting~895432393.deleting~957086549.deleting~943289565.deleting~124936451.deleting~004071477.deleting~750668574
app-logs.deleting~895432393.deleting~957086549.deleting~943289565.deleting~124936451.deleting~004071477.deleting~750668574.deleting~027954536
app-logs.deleting~895432393.deleting~957086549.deleting~943289565.deleting~124936451.deleting~016747332
app-logs.deleting~895432393.deleting~957086549.deleting~943289565.deleting~124936451.deleting~016747332.deleting~417902521
app-logs.deleting~895432393.deleting~957086549.deleting~943289565.deleting~124936451.deleting~016747332.deleting~874258893
app-logs.deleting~895432393.deleting~957086549.deleting~943289565.deleting~124936451.deleting~078643155
app-logs.deleting~895432393.deleting~957086549.deleting~943289565.deleting~124936451.deleting~078643155.deleting~691026306
^Capp-logs.deleting~895432393.deleting~957086549.deleting~943289565.deleting~124936451.deleting~078643155.deleting~691026306.deleting~677063412
```

My version is build from HEAD code.

**Comment by jasonrm** [comment]
I'm still testing various versions (1.5.2 with go 1.7 and 1.5.3 with go 1.8 so far), as I'm not sure why I don't see this issue always, but in case this helps someone else in the meantime…

When watching `journalctl -f -u kubelet`, it's just a steady stream of errors like,
```
55:30.3163065 +0000 UTC (durationBeforeRetry 500ms). Error: UnmountVolume.TearDown failed for volume "kubernetes.io/empty-dir/280de918-09a5-11e7-b475-001ec9399ff3-artfire-cache.deleting~210000594.deleting~029567820.deleting~766479260.deleting~071070870.deleting~580459210.deleting~589586053" (volume.spec.Name: "artfire-cache.deleting~210000594.deleting~029567820.deleting~766479260.deleting~071070870.deleting~580459210.deleting~589586053") pod "280de918-09a5-11e7-b475-001ec9399ff3" (UID: "280de918-09a5-11e7-b475-001ec9399ff3") with: rename /var/lib/kubelet/pods/280de918-09a5-11e7-b475-001ec9399ff3/volumes/kubernetes.io~empty-dir/artfire-cache.deleting~210000594.deleting~029567820.deleting~766479260.deleting~071070870.deleting~580459210.deleting~589586053 /var/lib/kubelet/pods/280de918-09a5-11e7-b475-001ec9399ff3/volumes/kubernetes.io~empty-dir/artfire-cache.deleting~210000594.deleting~029567820.deleting~766479260.deleting~071070870.deleting~580459210.deleting~589586053.deleting~650094748: file exists
```

Running `find /var/lib/kubelet -xdev -type d | wc -l` on the servers running kubelet shows there are an __insane__ number of these `deleting` files. The worst I've seen was 800k, and that was after letting an `rm -rf` run for minutes before checking to see what it was taking so long.

```
amber:
    118
atlanta:
    149
avenue:
    135
caramel:
    366
digital:
    93
flex:
    380806
remark:
    432337
```

When things go sideways, kubelet CPU and memory usage is crazy (compared to normal) and because the filesystem (ZFS in my case) is continually looking through 100k-1M+ of files/dirs load average spikes,

```
top - 21:54:09 up 1 day,  3:18,  2 users,  load average: 908.84, 677.34, 278.02
Tasks: 419 total,   1 running, 417 sleeping,   0 stopped,   1 zombie
%Cpu(s):  6.9 us,  0.4 sy,  0.0 ni, 91.5 id,  0.7 wa,  0.0 hi,  0.4 si,  0.0 st
GiB Mem :   15.661 total,    6.707 free,    8.399 used,    0.556 buff/cache
GiB Swap:    0.000 total,    0.000 free,    0.000 used.    6.940 avail Mem

  PID USER      PR  NI    VIRT    RES  %CPU %MEM     TIME+ S COMMAND
 3553 root      20   0 42.514g 3.851g 170.2 24.6 132:21.40 S /usr/bin/kubelet
25634 root      20   0   49.7m  28.0m   0.7  0.2  18:38.61 S /usr/bin/kube-proxy
31962 root      20   0 3228.2m  60.2m   0.7  0.4  11:54.14 S /usr/bin/dockerd
30559 root      20   0  785.5m   4.4m   0.0  0.0   9:00.19 S collectd
16052 root      20   0   49.0m  23.1m   0.0  0.1   3:37.51 S /nginx-ingress-controller
 1590 root      20   0  439.6m 296.5m   6.6  1.8   2:13.74 S /usr/lib/systemd/systemd-journald
    2 root      20   0    0.0m   0.0m   0.0  0.0   2:03.73 S [kthreadd]
```

### Versions

```
amber:
    4.9.11-1-ARCH
caramel:
    4.8.13-1-ARCH
remark:
    4.10.2-1-ARCH
flex:
    4.10.2-1-ARCH
atlanta:
    4.10.1-1-ARCH
avenue:
    4.8.13-1-ARCH
digital:
    4.10.1-1-ARCH
```

```
Server Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.2+08e0995", GitCommit:"6bafc4a0a5a4ccd4008ec5caed548b5e6fe8b00d", GitTreeState:"dirty", BuildDate:"2017-01-12T10:24:17Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}

amber:
    Client Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.2+08e0995", GitCommit:"6bafc4a0a5a4ccd4008ec5caed548b5e6fe8b00d", GitTreeState:"dirty", BuildDate:"2017-01-12T10:24:17Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
atlanta:
    Client Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.2+08e0995", GitCommit:"6bafc4a0a5a4ccd4008ec5caed548b5e6fe8b00d", GitTreeState:"dirty", BuildDate:"2017-01-12T10:24:17Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
avenue:
    Client Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.1+82450d0", GitCommit:"8e0c8bc8d898d2097bb460f4d957fa557b0bfc8a", GitTreeState:"dirty", BuildDate:"2016-12-20T13:32:23Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
caramel:
    Client Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.1+82450d0", GitCommit:"8e0c8bc8d898d2097bb460f4d957fa557b0bfc8a", GitTreeState:"dirty", BuildDate:"2016-12-20T13:32:23Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
digital:
    Client Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.2+08e0995", GitCommit:"6bafc4a0a5a4ccd4008ec5caed548b5e6fe8b00d", GitTreeState:"dirty", BuildDate:"2017-01-12T10:24:17Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
flex:
    Client Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.3+029c3a4", GitCommit:"291d51166d0a085d3be254d9063e4f4203d666dd", GitTreeState:"dirty", BuildDate:"2017-02-18T11:13:41Z", GoVersion:"go1.8", Compiler:"gc", Platform:"linux/amd64"}
remark:
    Client Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.3+029c3a4", GitCommit:"291d51166d0a085d3be254d9063e4f4203d666dd", GitTreeState:"dirty", BuildDate:"2017-02-18T11:13:41Z", GoVersion:"go1.8", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by msau42** [comment]
The original comment reported the unmount error as "device or resource busy", which is not the same issue as the "deleting~" files.  The issue with "deleting~" files seems to be caused by compiling with go1.8.

**Comment by dims** [comment]
Should have been fixed already in master (example #43878)

**Comment by dims** [comment]
/close

please reopen if needed.
